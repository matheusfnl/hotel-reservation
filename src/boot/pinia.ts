import { boot } from 'quasar/wrappers'
import { createPinia, PiniaVuePlugin } from 'pinia'

export default boot(({ app }) => {
  const pinia = createPinia()

  app.use(PiniaVuePlugin)
  app.use(pinia)
})
