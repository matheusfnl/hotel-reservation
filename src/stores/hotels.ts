import { defineStore } from 'pinia'

import type { HotelsByPlace, Hotel } from '@/types'

import hotelsMock from '@/mock/hotel.json'

interface FetchHotelsParams {
  placeId?: number | string
  page?: number
  name?: string
  sort?: string
}

const allPlaceHotels = [...hotelsMock] as HotelsByPlace[]

export const useHotelsStore = defineStore('hotels', {
  state: () => ({ hotels: [] as Hotel[] }),
  actions: {
    async fetchHotels({ placeId, page = 1, name, sort }: FetchHotelsParams = {}) {
      try {
        const filterHotels = (): Hotel[] => {
          let placeHotels = placeId
            ? allPlaceHotels.find((hotel) => hotel.placeId === +placeId)?.hotels
            : allPlaceHotels.flatMap((place) => place.hotels)

          if (!placeHotels) {
            return []
          }

          if (name) {
            placeHotels = placeHotels.filter((hotel) =>
              hotel.name.toLowerCase().includes(name.toLowerCase()),
            )
          }

          if (['high_price', 'low_price'].includes(sort as string)) {
            return sort === 'low_price'
              ? placeHotels?.sort((a, b) => +a.price - +b.price)
              : placeHotels?.sort((a, b) => +b.price - +a.price)
          }

          if (sort === 'best_rated') {
            return placeHotels?.sort((a, b) => +b.stars - +a.stars)
          }

          return placeHotels?.sort((a, b) =>
            a.favorite !== b.favorite ? +b.favorite - +a.favorite : +b.stars - +a.stars,
          )
        }

        await new Promise((resolve) => setTimeout(resolve, 2000))

        const placeHotels = filterHotels()
        const itemsIndex = (page - 1) * 10

        this.hotels = placeHotels ? placeHotels.slice(itemsIndex, itemsIndex + 10) : []
      } catch (error: unknown) {
        return error
      }
    },
  },
})
