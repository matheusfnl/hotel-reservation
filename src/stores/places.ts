import { defineStore } from 'pinia'

import { Place } from '@/models/Place'
import type { Places } from '@/types'

import placesMock from '@/mock/place.json'

const places = placesMock as Places[]

export const usePlacesStore = defineStore('places', {
  state: () => ({ places: [] as Place[] }),
  actions: {
    async fetchPlaces() {
      try {
        await new Promise((resolve) => setTimeout(resolve, 2000))

        this.places = places.map((place) => new Place(place))
      } catch (error: unknown) {
        return error
      }
    },
  },
})
