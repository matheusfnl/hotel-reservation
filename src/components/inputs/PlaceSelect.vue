<template>
  <q-select
    v-model="placeSearch"
    use-input
    hide-selected
    fill-input
    outlined
    :dense="dense"
    clearable
    hide-dropdown-icon
    input-debounce="0"
    :placeholder="$t('hotel.search.inputs.destination.placeholder')"
    :options="filteredPlaces"
    option-label="formattedLabel"
    @filter="filterPlaces"
    @update:model-value="$emit('update:modelValue', $event)"
  >
    <template v-slot:no-option>
      <q-item>
        <q-item-section class="text-grey">
          {{ $t('hotel.search.inputs.destination.empty') }}
        </q-item-section>
      </q-item>
    </template>
  </q-select>
</template>

<script lang="ts" setup>
import { ref, onMounted } from 'vue'
import { useRoute } from 'vue-router'

import type { FormattedPlaces } from '@/types'

import { usePlacesStore } from '@/stores/places'

const placeStore = usePlacesStore()

const route = useRoute()

const emit = defineEmits<{
  'update:modelValue': [value: FormattedPlaces | null]
}>()

defineProps({
  dense: {
    type: Boolean,
    default: true,
  },
})

const placeSearch = ref<FormattedPlaces | null>(null)
const filteredPlaces = ref<FormattedPlaces[]>([])

const filterPlaces = (val: string, update: (fn: () => void) => void) => {
  const formattedPlaces = placeStore.places.map((place) => ({
    formattedLabel: `${place.name}, ${place.state.shortname}`,
    ...place,
  }))

  update(() => {
    const search = val.toLowerCase()
    filteredPlaces.value = formattedPlaces.filter(
      (place) =>
        place.formattedLabel.toLowerCase().includes(search) ||
        place.state.name.toLowerCase().includes(search),
    )
  })
}

const setPlace = (placeId: string | number) => {
  const formattedPlaces = placeStore.places.map((place) => ({
    formattedLabel: `${place.name}, ${place.state.shortname}`,
    ...place,
  }))

  emit('update:modelValue', formattedPlaces.find((place) => +place.placeId === +placeId) || null)
}

onMounted(async () => {
  await placeStore.fetchPlaces()

  if (route.params.placeId) {
    setPlace(route.params.placeId as string)
  }
})

defineExpose({
  setPlace,
})
</script>
