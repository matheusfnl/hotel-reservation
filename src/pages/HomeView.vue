<template>
  <div class="overflow-hidden">
    <div class="circle-1" />
    <div class="circle-2" />
    <div class="circle-3" />

    <div class="absolute-center full-width q-pa-md flex justify-center">
      <Form v-slot="{ meta }" style="width: fit-content">
        <q-card class="bg-white rounded-borders-xl">
          <q-card-section class="q-pa-xl text-center">
            <h2 class="text-h3 text-primary text-weight-medium q-ma-none q-mb-lg">
              {{ $t('home.cta.title') }}
            </h2>

            <p class="text-h6 q-mx-xl text-center q-ma-none q-mb-lg">
              {{ $t('home.cta.description') }}
              <span class="text-primary">{{ $t('home.cta.stay') }}</span
              >,
              {{ $t('home.cta.with') }}
              <span class="text-primary">{{ $t('home.cta.comfort') }}</span>
              {{ $t('home.cta.and') }}
              <span class="text-primary">{{ $t('home.cta.practicality') }}</span
              >.
            </p>

            <Field
              v-slot="{ field, errorMessage, meta }"
              name="place"
              :value="placeSearch"
              rules="placeSelected"
            >
              <div>
                <PlaceSelect
                  :model-value="placeSearch"
                  @update:model-value="(val) => handlePlaceUpdate(val, field)"
                  @blur="field.onBlur"
                />

                <span v-if="errorMessage && meta.touched" class="text-negative text-caption">
                  {{
                    $t('validation.required', {
                      field: $t('hotel.search.inputs.destination.label').toLowerCase(),
                    })
                  }}
                </span>
              </div>
            </Field>

            <q-btn
              rounded
              no-caps
              color="primary"
              size="lg"
              class="q-mt-lg"
              :disabled="!meta.valid"
              style="min-width: 160px"
              @click.prevent="navigateToHotels"
            >
              {{ $t('home.cta.search') }}
            </q-btn>
          </q-card-section>
        </q-card>
      </Form>
    </div>
  </div>
</template>

<script lang="ts" setup>
import { ref } from 'vue'
import type { Places } from '@/types'
import { Form, Field, defineRule } from 'vee-validate'
import type { FieldBindingObject } from 'vee-validate'
import { useRouter } from 'vue-router'

import PlaceSelect from '@/components/inputs/PlaceSelect.vue'

const router = useRouter()

const placeSearch = ref<Places | null>(null)
const handlePlaceUpdate = (val: Places | null, field: FieldBindingObject) => {
  placeSearch.value = val
  field.onChange(val)
}

const navigateToHotels = async () => {
  if (placeSearch.value?.placeId) {
    await router.push({
      name: 'hotels.place',
      params: { placeId: placeSearch.value.placeId },
    })
  }
}

const validatePlace = (value: Places | null) => {
  return !!value?.placeId
}

defineRule('placeSelected', validatePlace)
</script>

<style lang="scss" scoped>
.circle {
  &-1 {
    position: fixed;
    width: 120vh;
    height: 120vh;
    background: $primary;
    border-radius: 30% 70% 70% 30% / 30% 30% 70% 70%;
    left: -40vh;
    top: 75vh;
    z-index: 0;
    opacity: 1;
  }

  &-2 {
    position: fixed;
    width: 300vh;
    height: 300vh;
    background: $primary;
    border-radius: 53% 47% 47% 53% / 45% 0% 0% 55%;
    right: -130vh;
    top: 45vh;
    z-index: 0;
    opacity: 0.6;
  }

  &-3 {
    position: fixed;
    width: 120vh;
    height: 120vh;
    background: $primary;
    border-radius: 42% 58% 58% 42% / 53% 43% 57% 47%;
    left: 10vh;
    top: -90vh;
    z-index: 0;
    opacity: 0.7;
  }
}

.form-container {
  box-shadow: 0 10px 20px rgba(0, 0, 0, 1);
}
</style>
