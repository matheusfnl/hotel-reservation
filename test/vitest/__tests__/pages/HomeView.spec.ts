import { describe, it, expect, beforeEach, vi } from 'vitest'
import { setActivePinia, createPinia } from 'pinia'
import { mount } from '@vue/test-utils'
import { QBtn, QSelect } from 'quasar'
import { nextTick } from 'vue'
import { createRouter, createWebHistory } from 'vue-router'

import { usePlacesStore } from '@/stores/places'

import HomeView from '@/pages/HomeView.vue'
import PlaceSelect from '@/components/inputs/PlaceSelect.vue'

const mockPlaces = [
  { name: 'São Paulo', state: { name: 'São Paulo', shortname: 'SP' }, placeId: 1 },
  { name: 'Rio de Janeiro', state: { name: 'Rio de Janeiro', shortname: 'RJ' }, placeId: 2 },
]

const router = createRouter({
  history: createWebHistory(),
  routes: [],
})

describe('HomeView', () => {
  beforeEach(() => {
    setActivePinia(createPinia())
    const placesStore = usePlacesStore()
    placesStore.places = mockPlaces

    vi.spyOn(router, 'push').mockImplementation(() => Promise.resolve())
  })

  const mountWrapper = () => {
    return mount(HomeView, {
      global: {
        plugins: [router],
      },
    })
  }

  it('should render correctly', () => {
    const wrapper = mountWrapper()
    const html = wrapper
      .html()
      .replace(/id="f_[^"]+"/g, 'id="static-id"')
      .replace(/for="f_[^"]+"/g, 'for="static-for"')
      .replace(/aria-controls="f_[^"]+"/g, 'aria-controls="static-aria-controls"')

    expect(html).toMatchSnapshot()
  })

  it('should emit update:modelValue with selected place', async () => {
    const wrapper = mountWrapper()
    const selectComponent = wrapper.findComponent(PlaceSelect)
    const select = selectComponent.findComponent(QSelect)

    await select.vm.$emit('update:model-value', mockPlaces[0])
    await nextTick()

    expect(selectComponent.emitted('update:modelValue')).toBeTruthy()
    expect(selectComponent.emitted('update:modelValue')![0]).toEqual([
      {
        ...mockPlaces[0],
        formattedLabel: 'São Paulo, SP',
      },
    ])
  })

  it('should navigate to hotels on button click', async () => {
    const wrapper = mountWrapper()
    const selectComponent = wrapper.findComponent(PlaceSelect)
    const select = selectComponent.findComponent(QSelect)
    const button = wrapper.findComponent(QBtn)

    await select.vm.$emit('update:model-value', mockPlaces[0])
    await nextTick()

    await button.trigger('click')

    // eslint-disable-next-line @typescript-eslint/unbound-method
    expect(router.push).toHaveBeenCalledWith({
      name: 'hotels.place',
      params: { placeId: mockPlaces[0]!.placeId },
    })
  })
})
