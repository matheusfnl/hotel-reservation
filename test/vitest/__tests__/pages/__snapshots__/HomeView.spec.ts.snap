// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HomeView > should render correctly 1`] = `
"<div data-v-6fbf89c5="" class="overflow-hidden">
  <div data-v-6fbf89c5="" class="circle-1"></div>
  <div data-v-6fbf89c5="" class="circle-2"></div>
  <div data-v-6fbf89c5="" class="circle-3"></div>
  <div data-v-6fbf89c5="" class="absolute-center full-width q-pa-md flex justify-center">
    <form data-v-6fbf89c5="" novalidate="" style="width: fit-content;">
      <div data-v-6fbf89c5="" class="q-card bg-white rounded-borders-xl">
        <div data-v-6fbf89c5="" class="q-card__section q-card__section--vert q-pa-xl text-center">
          <h2 data-v-6fbf89c5="" class="text-h3 text-primary text-weight-medium q-ma-none q-mb-lg">Encontre o hotel perfeito para sua viagem</h2>
          <p data-v-6fbf89c5="" class="text-h6 q-mx-xl text-center q-ma-none q-mb-lg">Procure seu destino e encontre o hotel ideal para sua <span data-v-6fbf89c5="" class="text-primary">estadia</span>, com <span data-v-6fbf89c5="" class="text-primary">conforto</span> e <span data-v-6fbf89c5="" class="text-primary">praticidade</span>. </p>
          <div data-v-6fbf89c5=""><label data-v-cf369733="" data-v-6fbf89c5="" class="q-field row no-wrap items-start q-field--outlined q-select q-field--auto-height q-select--with-input q-select--without-chips q-select--single q-field--dense place-select" for="static-for">
              <!---->
              <div class="q-field__inner relative-position col self-stretch">
                <div class="q-field__control relative-position row no-wrap" tabindex="-1">
                  <div class="q-field__control-container col relative-position row no-wrap q-anchor--skip">
                    <div class="q-field__native row items-center"><input class="q-field__input q-placeholder col" type="search" tabindex="0" role="combobox" aria-readonly="false" aria-autocomplete="list" aria-expanded="false" aria-controls="static-aria-controls" placeholder="Cidade, UF" selected-item-label="selectedLabel" id="static-id" value=""></div>
                    <!---->
                  </div>
                  <!---->
                  <!---->
                </div>
                <!---->
              </div>
              <!---->
            </label>
            <!--v-if-->
          </div><button data-v-6fbf89c5="" class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--rectangle q-btn--rounded bg-primary text-white q-btn--actionable q-focusable q-hoverable q-btn--no-uppercase q-mt-lg" style="font-size: 20px; min-width: 160px;" tabindex="0" type="button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row">Procurar</span></button>
        </div>
      </div>
    </form>
  </div>
</div>"
`;
